src/dbnode/storage/index/block.go

```diff
@@ -162,32 +471,242 @@ func (b *block) WriteBatch(inserts *WriteBatch) (WriteBatchResult, error) {
-		// Avoid marking these as success
+		// Avoid marking these as success.
 		inserts.MarkUnmarkedEntryError(err.Err, err.Idx)
 	}
 
-	// mark all non-error inserts success, so we don't repeatedly index them
+	// Mark all non-error inserts success, so we don't repeatedly index them.
 	inserts.MarkUnmarkedEntriesSuccess()
 	return WriteBatchResult{
 		NumSuccess: int64(inserts.Len() - numErr),
 		NumError:   int64(numErr),
 	}, partialErr
 }
 
-func (b *block) executorWithRLock() (search.Executor, error) {
-	var expectedReaders int
-	if b.activeSegment != nil {
-		expectedReaders++
+func (b *block) foregroundCompactWithBuilder(builder segment.DocumentsBuilder) error {
+	// We inserted some documents, need to compact immediately into a
+	// foreground segment.
+	b.Lock()
+	foregroundSegments := b.foregroundSegments
+	b.Unlock()
+
+	segs := make([]compaction.Segment, 0, len(foregroundSegments)+1)
+	segs = append(segs, compaction.Segment{
+		Age:  0,
+		Size: int64(len(builder.Docs())),
+		Type: segments.MutableType,
+	})
+	for _, seg := range foregroundSegments {
+		segs = append(segs, compaction.Segment{
+			Age:     seg.Age(),
+			Size:    seg.Segment().Size(),
+			Type:    segments.FSTType,
+			Segment: seg.Segment(),
+		})
 	}
+
+	plan, err := compaction.NewPlan(segs, b.opts.ForegroundCompactionPlannerOptions())
+	if err != nil {
+		return err
+	}
+
+	// Check plan.
+	var planErr error
+	switch {
+	case len(builder.Docs()) == 0:
+		// Must be compacting with a builder that has some inserted documents.
+		planErr = errForegroundCompactorNoBuilderDocs
+	case len(plan.Tasks) == 0:
+		// Must generate a plan since we gave it a mutable segment type to compact.
+		planErr = errForegroundCompactorNoPlan
+	case !taskHasNilSegment(plan.Tasks[0]):
+		// First task of plan must include the builder, so we can avoid resetting it
+		// for the first task, but then safely reset it in consequent tasks.
+		planErr = errForegroundCompactorBadPlanFirstTask
+	}
+	if planErr != nil {
+		return planErr
+	}
+
+	// Move any unused segments to the background.
+	b.Lock()
+	b.maybeMoveForegroundSegmentsToBackgroundWithLock(plan.UnusedSegments)
+	b.Unlock()
+
+	n := b.compactionsForeground
+	b.compactionsForeground++
+
+	logger := b.logger.WithFields(
+		xlog.NewField("block", b.startTime.String()),
+		xlog.NewField("numForegroundCompaction", n),
+	)
+	log := n%compactDebugLogEvery == 0
+	if log {
+		for i, task := range plan.Tasks {
+			summary := task.Summary()
+			logger.WithFields(
+				xlog.NewField("task", i),
+				xlog.NewField("numMutable", summary.NumMutable),
+				xlog.NewField("numFST", summary.NumFST),
+				xlog.NewField("cumulativeMutableAge", summary.CumulativeMutableAge.String()),
+				xlog.NewField("cumulativeSize", summary.CumulativeSize),
+			).Debug("planned foreground compaction task")
+		}
+	}
+
+	// Run the plan.
+	sw := b.metrics.foregroundCompactionPlanRunLatency.Start()
+	defer sw.Stop()
+
+	// Run the first task, without resetting the builder.
```

Comment:
is resetting the builder expensive?

Reply:
So this is confusing, but basically the first task is required to always include the builder (because its a mutable segment so must be compacted immediately by the settings we specify for the planner for foreground compaction), that's why I do the sanity check making sure the first task is the `nil` Segment that signifies it represents the builder.  

The builder already has a ton of documents from the `WriteBatch(...)` call, hence must not be reset, otherwise the documents it has doesn't get compacted into a new segment or with another segment.